src/java.base/share/classes/java/util/Arrays.java

```diff
@@ -1243,7 +1245,8 @@ private static void swap(@UnknownSignedness Object[] x, int a, int b) {
      * @throws IllegalArgumentException (optional) if the comparator is
      *         found to violate the {@link Comparator} contract
      */
-    public static <T> void sort(@PolyNull @UnknownSignedness T[] a, @Nullable Comparator<? super T> c) {
+    @CFComment("A particular comparator could allow null elements in the array. In general null is forbidden.")
```

Comment:
Despite this, I think that the old type-parameter declaration `<T>` is OK: If the caller passes a `Comparator<@NonNull Foo>`, then the caller will also have to pass an array of non-nullable foos.

Reply:
Sorry for dropping this.

Yes, you have convinced me: I had not been taking into account that the comparator parameter was `@Nullable`.

I guess, with a `@Nullable` comparator, you may have the additional problem that you might pass values that aren't `Comparable` at all :) But our job with the nullness annotations is just to protect from the nullness errors, so let's do that.

I have no reason to believe that this signature will cause us trouble inside Google, but I'll let you know (and maybe just modify the signature internally :)) if I see more than a handful of errors as a result.

Reply:
Oh, sorry, you'd asked about the comment text: "We require non-nullable array elements because `sort` calls `compareTo` on them if the caller omits the Comparator?" I don't know, something like that?

Now that I write that, I realize that there's an alternative, which is to refuse to make the _comparator_ parameter `@Nullable`, even though users can successfully pass `null` :) And actually, wait: Did you get your wish? I see `sort(Object[])`, which users could always use if they want natural ordering. (And we're already fixing its annotations in this PR.) I take your point that the current 2-arg method _can_ sometimes run with a null comparator, but it _can_ sometimes run with null array elements, too. So maybe we get to pick which of those to support without producing warnings?

I do still see a couple callers in third-party code that call `Arrays.sort(..., null)`, and there may be others that my quick search missed. But we would very possibly not break any of our users of eisop. Then again, as I said above, we might not break any of our users of eisop if we go the other way, either :)

Reply:
@cpovirk  Thanks for your comments! I've changed this to be more null-hostile. There are equivalent versions of these methods without a comparator and people should either use that other version or suppress the warning.
If tests pass I'll merge this. Please do let me know if this is causing too many false positives and we can revisit.

Reply:
We just tested this in Google. There were two failures:
- one that was testing a custom implementation of `java.util.Arrays`, which naturally needs to pass `null` in order to test that case
- one that was passing `foo ? reverseOrder() : null`, which could switch to `foo ? reverseOrder() : naturalOrder()` or which could use `if (foo) { sort(x, reverseOrder()); } else { sort(x) }`.

(And for Kotlin, etc., we're using the more permissive JSpecify stubs, so there's no danger there.)